File: prak-4/responsi/1/Kompleks.java
public class Kompleks {

    private static int n_kompleks = 0;
    private int real;
    private int imaginer;

    // ctor tanpa parameter
    // inisialisasi seluruh koefisien dengan nilai 0
    public Kompleks() {
        this(0, 0);
    }

    // ctor dengan parameter
    public Kompleks(int real, int imaginer) {
        this.real = real;
        this.imaginer = imaginer;  
        n_kompleks ++;
    }
    
    //mengembalikan bagian riil
    public int getReal() {
        return real;
    }

    // mengembalikan bagian imaginer
    public int getImaginer() {
        return imaginer;
    }

    // mengisi bagian riil
    public void setReal(int real) {
        this.real = real;
    }

    // mengisi bagian imaginer
    public void setImaginer(int imaginer) {
        this.imaginer = imaginer;
    }

    // operator+ untuk melakukan penjumlahan dengan rumus berikut
    public Kompleks plus(Kompleks b) {
        Kompleks res = new Kompleks();

        res.setReal(getReal() + b.getReal());
        res.setImaginer(getImaginer() + b.getImaginer());

        return res;
    }

    // operator- untuk melakukan pengurangan dengan rumus berikut
    public Kompleks minus(Kompleks b) {
        Kompleks res = new Kompleks();

        res.setReal(getReal() - b.getReal());
        res.setImaginer(getImaginer() - b.getImaginer());

        return res;
    }

    // operator* untuk melakukan perkalian dengan rumus berikut
    public Kompleks multiply(Kompleks b) {
        Kompleks res = new Kompleks();

        res.setReal((getReal() * b.getReal()) - (getImaginer() * b.getImaginer()));
        res.setImaginer((getImaginer() * b.getReal()) + (getReal() * b.getImaginer()));

        return res;
    }

    // operator* untuk mengkalikan bilangan kompleks dengan konstanta
    public Kompleks multiply (int x) {
        Kompleks res = new Kompleks();

        res.setReal(getReal() * x);
        res.setImaginer(getImaginer() * x);

        return res;
    }

    // mengembalikan jumlah instance yang pernah dibuat
    public static int countKompleksInstance() {
        return n_kompleks;
    }

    // mencetak bilangan kompleks ke layar, diakhiri dengan end-of-line 
    // contoh:
    // 3+5i
    // 0
    // 3i
    // -3
    // -5-4i
    public void print() {
        if (real == 0 && imaginer == 0) {
            System.out.println(0);
        } else if (real == 0) {
            System.out.println(imaginer + "i");
        } else if (imaginer == 0) {
            System.out.println(real);
        } else {
            System.out.print(real);
        
            if(imaginer > 0) {
                System.out.print("+");
            }

            System.out.println(imaginer + "i");
        }
    } 

    public static void main(String[] args) {
        Kompleks k = new Kompleks(2, -1);

        k.print();

        System.out.println(Kompleks.n_kompleks);
        System.out.println(k.countKompleksInstance());

        Kompleks k2 = new Kompleks(5, 3);

        k = k.multiply(k2);

        k.print();
    }
}

File: prak-1/prak/3/Kompleks.cpp
#include <iostream>
#include "Kompleks.h"
using namespace std;

#define KK Kompleks

int KK::n_kompleks = 0;

// ctor tanpa parameter
// inisialisasi seluruh koefisien dengan nilai 0
KK::Kompleks() : Kompleks(0, 0) {

}

// ctor dengan parameter
KK::Kompleks(int real, int imaginer) {
	this->real = real;
	this->imaginer = imaginer;
	KK::n_kompleks++;
}

//mengembalikan bagian riil
int KK::GetReal() const {
	return this->real;
}

// mengembalikan bagian imaginer
int KK::GetImaginer() const {
	return this->imaginer;
}

// mengisi bagian riil
void KK::SetReal(int x) {
	this->real = x;
}

// mengisi bagian imaginer
void KK::SetImaginer(int y) {
	this->imaginer = y;
}

// operator overloading

// operator+ untuk melakukan penjumlahan dengan rumus berikut
// (a + bi) + (c + di) = (a+c) + (b+d)i
Kompleks operator+ (const Kompleks& a, const Kompleks& b) {
	KK res;

	res.real = (a.GetReal() + b.GetReal());
	res.imaginer = (a.GetImaginer() + b.GetImaginer());

	return res;
}

// operator- untuk melakukan pengurangan dengan rumus berikut
// (a + bi) − (c + di) = (a−c) + (b−d)i
Kompleks operator- (const Kompleks& a, const Kompleks& b) {
	KK res;

	res.real = (a.GetReal() - b.GetReal());
	res.imaginer = (a.GetImaginer() - b.GetImaginer());

	return res;
}

// operator* untuk melakukan perkalian dengan rumus berikut
// (a + bi)(c + di) = ac + bci + adi + bd i^2 = (ac−bd) + (bc+ad)i
Kompleks operator* (const Kompleks& a, const Kompleks& b) {
	KK res;

	res.real = ((a.GetReal() * b.GetReal()) - (a.GetImaginer() * b.GetImaginer()));
	res.imaginer = ((a.GetImaginer() * b.GetReal()) + (a.GetReal() * b.GetImaginer()));

	return res;
}

// operator* untuk mengkalikan bilangan kompleks dengan konstanta
// (a + bi)(c) = (ac) + (bc)i
Kompleks operator* (const Kompleks& a, const int x) {
	KK res;

	res.real = (a.GetReal() * x);
	res.imaginer = (a.GetImaginer() * x);

	return res;
}

// operator* untuk mengkalikan bilangan kompleks dengan konstanta (sifat komutatif)
Kompleks operator* (const int x, const Kompleks& a) {
	KK res(a);

	return res * x;
}

// mengembalikan jumlah instance yang pernah dibuat
int KK::CountKompleksInstance() {
	return KK::n_kompleks;
}

// mencetak bilangan kompleks ke layar, diakhiri dengan end-of-line 
// contoh:
// 3+5i
// 0+0i
// -5-4i
void KK::Print() {
	cout << this->real;
	if (this->imaginer >= 0) {
		cout << "+";
	}
	cout << this->imaginer << "i" << endl;
}